BUGFIX: Correctly reflect builtin types in method arguments

albe opened the pull request:
Before builtin types like `object` or `iterable` in method arguments would lead to being reflected as if they were class types, leading to the AOP proxy builder prefixing those types with `\` which results in a PHP error "Type declaration 'object' must be unqualified". Only "simple" types were handled correctly as they had a separate check.
This adds a test that verifies that builtin types like `object` and `iterable` are correctly reflected as non-class types in the ReflectionService.

albe committed:
TASK: Add test to verify builtin types are properly reflected in method arguments

albe committed:
BUGFIX: Correctly reflect builtin types for method arguments

albe committed:
BUGFIX: Correctly reflect builtin types for method arguments

kdambekalns committed:
Merge branch '7.0' into fix-builtin-parameter

kdambekalns committed:
TASK: Code clean up

Does some small-scale code cleanup to parts I came across.

kdambekalns commented:
Ok, the change subtly changes the way "simple types" are reflected on method parameters… A "simple type", e.g. a string was not declared "scalar" before. In combination with the AOP proxy building, this now leads to type declarations being stricter than in the original class.

The reason is we now enter this branch:

https://github.com/neos/flow-development-collection/blob/f8aec784fd283e1e9fce707ee5032b3f924f8b59/Neos.Flow/Classes/ObjectManagement/Proxy/ProxyMethod.php#L275-L276

which leads to e.g. `?string` being added in this case from `Session`:

`Fatal error:  Declaration of Neos\Flow\Session\Session::destroy(?string $reason = NULL) must be compatible with Neos\Flow\Session\Session_Original::destroy($reason = NULL) in /…/Flow_Object_Classes/Neos_Flow_Session_Session.php on line 1064`

kdambekalns committed:
BUGFIX: Fix broken AOP proxy building

albe commented:
> A "simple type", e.g. a string was **not** declared "scalar" before.

🤯 I totally misread the code there

kdambekalns commented:
> > A "simple type", e.g. a string was **not** declared "scalar" before.
> 
> 🤯 I totally misread the code there

To be fair, that scalar types are not makred as such is… counter-intuitive, at least. 🙃

kdambekalns committed:
Merge branch '7.0' into fix-builtin-parameter

albe commented:
Should we maybe adapt https://github.com/neos/flow-development-collection/blob/c4df7ec2520fc8cf85ae15c3570e0483d8c3782b/Neos.Flow/Classes/Reflection/ReflectionService.php#L1384-L1386 as well in this PR to keep the two places more in line? Note though that return types have `self` and `static`, which are **NOT** "builtin" (for whatever obscure reason, `void` is builtin though): https://3v4l.org/gCd1H#v8.0.17

kdambekalns committed:
Merge branch '7.0' into fix-builtin-parameter

kdambekalns reviewed (approved):
Count as half of an approval, as I was involved in tweaking the code…

albe committed:
TASK: Adjust return type full qualification check for all builtins

Also checks for non-builtin return type declarations `self`, `static` and `parent`.

albe commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1381,7 +1381,7 @@ protected function reflectClassMethod($className, MethodReflection $method)
 
         $returnType = $method->getDeclaredReturnType();
-        if ($returnType !== null && !in_array($returnType, ['self', 'null', 'callable', 'void', 'iterable', 'object', 'mixed']) && !TypeHandling::isSimpleType($returnType)) {
+        if ($returnType !== null && !in_array($returnType, ['self', 'static', 'parent']) && !$method->getReturnType()->isBuiltin()) {
```
See https://github.com/neos/flow-development-collection/pull/2819 for a fix for return types in 6.3 without the builtin check

albe commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1381,7 +1381,7 @@ protected function reflectClassMethod($className, MethodReflection $method)
 
         $returnType = $method->getDeclaredReturnType();
-        if ($returnType !== null && !in_array($returnType, ['self', 'null', 'callable', 'void', 'iterable', 'object', 'mixed']) && !TypeHandling::isSimpleType($returnType)) {
+        if ($returnType !== null && !in_array($returnType, ['self', 'static', 'parent']) && !$method->getReturnType()->isBuiltin()) {
```
```suggestion
        if ($returnType !== null && !in_array($returnType, ['self', 'static', 'parent']) && $method->getReturnType() !== null && !$method->getReturnType()->isBuiltin()) {
```

albe committed:
TASK: Check for null to make psalm happy

albe committed:
Merge branch '7.0' into fix-builtin-parameter

mhsdesign commented on Neos.Flow/Classes/ObjectManagement/Proxy/ProxyMethod.php:
```diff
@@ -368,7 +368,8 @@ protected function getMethodVisibilityString()
             return 'protected';
-        } elseif ($this->reflectionService->isMethodPrivate($this->fullOriginalClassName, $this->methodName)) {
+        }
+        if ($this->reflectionService->isMethodPrivate($this->fullOriginalClassName, $this->methodName)) {
```
hehe good one xD

mhsdesign commented:
I dont understand why we have this and #2819 xD

albe commented:
> I dont understand why we have this and #2819 xD

tbh. I don't 100% know any more, but AFAIS the 6.3 PR only cares about the 'parent'/'static' cases and was supposed to be a quick fix for this common case, while this handles the more generic "isBuiltinType()" case, which could only be fixed properly in 7.0+

mhsdesign commented:
I think we wont need this fix for Flow 9 and even 8.3 seems to work fine?
